### PR TITLE
Add unit tests for IArrayCreationExpression and IArrayInitializer

### LINKED ISSUE #17596: [IExpression] IOperation unit tests for array related expressions
- [ ] IArrayCreationExpression
- [ ] IArrayInitializer

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #17596

Issues identified while adding unit tests: https://github.com/dotnet/roslyn/issues/18150

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @dotnet/analyzer-ioperation @AlekseyTs @cston for review.

**Comment by genlu** [comment]
Do we want to test erroneous code as well?

**Comment by mavasani** [comment]
@genlu I have added a unit test for invalid array creation in this PR: https://github.com/dotnet/roslyn/pull/17938/files#diff-b955da5b79c539293b33f4d4a23510deR336

I will add some more invalid creation tests.

**Comment by mavasani** [comment]
@genlu Added unit tests for error cases.

**Comment by mavasani** [comment]
@heejaechang @333fred @jinujoseph - can you please take a look?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+        public void SimpleArrayCreation_DimensionWithImplicitConversion()
+        {
+            string source = @"
+class M { }
+
+class C
+{
+    public void F(char dimension)
+    {
+        var a = /*<bind>*/new M[dimension]/*</bind>*/;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: Can you also include some error tests with dimensions that are not implicitly convertible, but are used as implicitly convertible anyway?

Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: Also tests for the result being a method call, with no conversion, correct/incorrect implicit conversion, correct/incorrect explicit conversion?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+        [Fact, WorkItem(17596, "https://github.com/dotnet/roslyn/issues/17596")]
+        public void ArrayCreationWithInitializer_MultipleInitializersWithConversions()
+        {
+            string source = @"
+class C
+{
+    public void F()
+    {
+        var a = """";
+        var b = /*<bind>*/new[] { ""hello"", a, null }/*</bind>*/;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: Could you also include tests for arrays created with explicit sizes and content initializers, correct and incorrect?

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 3]
```diff
@@ -0,0 +1,435 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: Do we need to add `ReDim` tests here?

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+    Public Sub F()
+        Const dimension As Integer = 1
+        Dim a = New M(dimension) {}'BIND:"New M(dimension) {}"
+    End Sub
+End Class
+    ]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 1, Element Type: M) (OperationKind.ArrayCreationExpression, Type: M())
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 2)
```
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: Why in the heck is this an `IBinaryExpression`?!

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+Class C
+    Public Sub F(dimension As Integer)
+        Dim a = New M(dimension) {}'BIND:"New M(dimension) {}"
+    End Sub
+End Class
+    ]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 1, Element Type: M) (OperationKind.ArrayCreationExpression, Type: M())
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32)
```
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: Again with the `IBinaryExpression`

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+    Public Sub F()
+        Dim b As Byte(,,) = New Byte(0, 1, 2) {}'BIND:"New Byte(0, 1, 2) {}"
+    End Sub
+End Class
+
+    ]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 3, Element Type: System.Byte) (OperationKind.ArrayCreationExpression, Type: System.Byte(,,))
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 1)
```
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: Lots of `IBinaryOperationExpression`s.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+Class C
+    Public Sub F()
+        Dim a As Integer()(,) = New Integer(0)(,) {}'BIND:"New Integer(0)(,) {}"
+    End Sub
+End Class
+    ]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 1, Element Type: System.Int32(,)) (OperationKind.ArrayCreationExpression, Type: System.Int32()(,))
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 1)
```
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: Again

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+Class C
+    Public Sub F()
+        Dim a = New String(1,) {}'BIND:"New String(1,) {}"
+    End Sub
+End Class
+    ]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 2, Element Type: System.String) (OperationKind.ArrayCreationExpression, Type: System.String(,), IsInvalid)
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 2)
```
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: And again. In this case, the resulting IOp tree here is very different from the C# version. Is this another result of the highest/lowest difference?

**Review by 333fred — DISMISSED** [review]
Most of the C# feedback applies to VB as well

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: That is how the language works, if you specified 1, the size of the allocated array will be two (one is added). I believe this is for compatibility with VB6, which had one-based arrays. So, it was common to have VB6 code to allocate an array of size N and then index into the array using N as index.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+class C
+{
+    public void F()
+    {
+        var a = /*<bind>*/new string[1]/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+IArrayCreationExpression (Dimension sizes: 1, Element Type: System.String) (OperationKind.ArrayCreationExpression, Type: System.String[])
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: >System.String [](start = 60, length = 13)

A question for design team, is there a value in duplicating the element type in a special property, can't we get it from the type for the node?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+IArrayCreationExpression (Dimension sizes: 1, Element Type: System.String) (OperationKind.ArrayCreationExpression, Type: System.String[], IsInvalid)
+  ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
+  IArrayInitializer (OperationKind.ArrayInitializer, IsInvalid)
+    IConversionExpression (ConversionKind.Invalid, Implicit) (OperationKind.ConversionExpression, Type: System.String, IsInvalid)
+      ILiteralExpression (Text: 1) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
+";
+            VerifyOperationTreeForTest<ArrayCreationExpressionSyntax>(source, expectedOperationTree);
+        }
+    }
+}
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: Consider also adding a test for array creation of the form ```int[] x = { 1, 2 };```.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: I would test ReDim statement separately, I am not sure if we even have representation for it in the current IOperation model.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs (hunk 10 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: Consider also testing 
```
new[] ;
new[2] ;
```
with and without target type.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+    End Sub
+End Class
+    ]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 1, Element Type: M) (OperationKind.ArrayCreationExpression, Type: M())
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 2)
+    Left: ILocalReferenceExpression: dimension (OperationKind.LocalReferenceExpression, Type: System.Int32, Constant: 1)
+    Right: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
+  IArrayInitializer (OperationKind.ArrayInitializer)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: >IArrayInitializer (OperationKind.ArrayInitializer) [](start = 2, length = 50)

It feels like there should be no initializer for case like this.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+    End Sub
+End Class
+    ]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 1, Element Type: M) (OperationKind.ArrayCreationExpression, Type: M())
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32)
+    Left: IParameterReferenceExpression: dimension (OperationKind.ParameterReferenceExpression, Type: System.Int32)
+    Right: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
+  IArrayInitializer (OperationKind.ArrayInitializer)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: >IArrayInitializer (OperationKind.ArrayInitializer) [](start = 2, length = 50)

It feels like there should be no initializer for a case like this. I see this as a case where presence of ```{}``` is just a syntax requirement, in fact it is an error to have it non-empty.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+
+            VerifyOperationTreeForTest(Of ArrayCreationExpressionSyntax)(source, expectedOperationTree)
+        End Sub
+
+        <Fact, WorkItem(17596, "https://github.com/dotnet/roslyn/issues/17596")>
+        Public Sub ArrayCreationOfSingleDimensionalArrays()
+            Dim source = <![CDATA[
+Class C
+    Public Sub F()
+        Dim a = {{1, 2, 3}, {4, 5, 6}}'BIND:"{{1, 2, 3}, {4, 5, 6}}"
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: >{{1, 2, 3}, {4, 5, 6}} [](start = 16, length = 22)

Do we have tests for erroneous array literals (mismatched element types, mismatched dimensions, etc.)?

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+IArrayCreationExpression (Dimension sizes: 1, Element Type: C) (OperationKind.ArrayCreationExpression, Type: C(), IsInvalid)
+  ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
+  IArrayInitializer (OperationKind.ArrayInitializer, IsInvalid)
+    IConversionExpression (ConversionKind.Basic, Implicit) (OperationKind.ConversionExpression, Type: C, IsInvalid)
+      ILiteralExpression (Text: 1) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
+]]>.Value
+
+            VerifyOperationTreeForTest(Of ArrayCreationExpressionSyntax)(source, expectedOperationTree)
+        End Sub
+    End Class
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: >End Class [](start = 4, length = 9)

Consider testing cases like these:
```
        x = New Integer(2) {1, 2}
        x = New Integer(x(0)) {1, 2}

        Dim y = New Integer(x(0), x(1)) {}
        y = New Integer(1, 1) {}
        y = New Integer(1, 1) {{}}
        y = New Integer(1, 1) {{}, {}}
        y = New Integer(1, 1) {{1, 2}}
        y = New Integer(1, 1) {{1, 2}, {}}
        y = New Integer(1, 1) {{1, 2}, {1, 2}}
```

**Comment by AlekseyTs** [comment]
How should the following array creation be exposed?
```
Dim x(2) As Integer
```

Perhaps it should be covered variable declarations. CC @333fred

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs (hunk 9 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: This PR went dormant and I forgot to file an issue for this. Opened https://github.com/dotnet/roslyn/issues/22265

**Comment by mavasani** [comment]
@333fred @AlekseyTs I have addressed your feedback and the PR is ready for review.

One additional API change has been done: I have removed `IArrayCreationExpression.ElementType` as per https://github.com/dotnet/roslyn/issues/22265

**Comment by mavasani** [comment]
@dotnet-bot retest ubuntu_14_debug_prtest please

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs [hunk 15]
```diff
@@ -0,0 +1,853 @@
...
+class C
+{
+    public void F(int dimension)
+    {
+        var x = /*<bind>*/new[2]/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.Int32[], IsInvalid) (Syntax: 'new[2]/*</bind>*/')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: >System.Int32[] [](start = 71, length = 14)

It feels strange that we decided the type the array is int[]. Please open an issue to investigate.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb (hunk 14 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: Are we testing these scenarios?

---
In reply to: [108760949](https://github.com/dotnet/roslyn/pull/18154#discussion_r108760949) [](ancestors = 108760949)

**Comment by AlekseyTs** [comment]
We should have a test for this declaration ```Dim x(2) As Integer```

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 8)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs (hunk 15 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: https://github.com/dotnet/roslyn/issues/22315

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb (hunk 14 above)
Comment (mavasani) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb]: Done - sorry I had missed this comment.

**Comment by mavasani** [comment]
@AlekseyTs - added requested unit tests.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 11)

**Comment by mavasani** [comment]
@cston @jcouv @dotnet/roslyn-compiler Can I have one more review please?

**Review by cston — APPROVED** [review]

**Pull request merged**